internal/server/admin/certificate_service.go

```diff
@@ -35,6 +82,11 @@ func (s *CertificateService) GetCertificate(ctx context.Context,
 	if err != nil {
 		return nil, s.err(ctx, err)
 	}
+	if !util.IsReference(result.Certificate.Cert) {
+		if err := extractCertInfo(result.Certificate); err != nil {
```

Comment:
Instead of repeating this logic over and over, let's just implement the [model.ObjectWithResourceDTO](https://github.com/Kong/koko/blob/v0.1.0/internal/model/object.go#L43) interface, have it use the default `MarshalResourceJSON()` behavior, and implement its own `UnmarshalResourceJSON()` method that uses the default logic, but handles this cert extraction.

This way, the implementation will be clean and we don't need to litter the service-level code with this.

Reply:
Very good point, I forgot we have such tooling already in place!

I tried to implement this, but it's not super clear to me how I'm expected to do so. I added something like the following, but I think I'm missing something since the receiving `r.Certificate` is always empty:

```

func (r Certificate) MarshalResourceJSON() ([]byte, error) {
	return json.Marshal(r)
}

func (r Certificate) UnmarshalResourceJSON([]byte) error {
	if !isReference(r.Certificate.Cert) {
		if err := extractCertInfo(r.Certificate); err != nil {
			return err
		}
	}
	return json.Unmarshal([]byte(fmt.Sprintf("%v", r)), &v1.Certificate{})
}
```

Reply:
You'll need to call unmarshal first and then extract the info on the certificate. 😄